taichi/ui/common/scene_base.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <vector>
+#include "taichi/ui/common/field_info.h"
+#include "taichi/ui/common/renderable_info.h"
+#include "taichi/ui/common/camera.h"
+#include "taichi/ui/utils/utils.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+struct PointLight {
+  glm::vec3 pos;
+  glm::vec3 color;
+};
+
+struct MeshInfo {
+  RenderableInfo renderable_info;
+  glm::vec3 color;
+  float shininess;
+};
+
+struct ParticlesInfo {
+  RenderableInfo renderable_info;
+  glm::vec3 color;
+  float radius;
+  float shininess;
+};
+
+#define MAX_POINTLIGHTS 16
```

Comment:
```suggestion
constexpr int kMaxPointLights = 16;
```

Reply:
Probably don't need this max values in the future, deferred lighting easily deals with this, and in forward we can totally support a variable length array.